ring/token_range_test.go

```diff
@@ -0,0 +1,306 @@
+			for z := 1; z < numZones; z++ {
+				for i := 0; i < len(testData.zoneTokens); i++ {
+					id := fmt.Sprintf("instance-%d-%d", z, i)
+					tokens := GenerateTokens(512, allTokens)
+					instances[id] = generateRingInstanceWithInfo(id, fmt.Sprintf("zone-%d", z), tokens, time.Now())
+					allTokens = append(allTokens, tokens...)
+				}
+			}
+
+			// Initialise the ring.
+			ringDesc := &Desc{Ingesters: instances}
+			ring := Ring{
+				cfg:                  Config{HeartbeatTimeout: time.Hour, ZoneAwarenessEnabled: true, SubringCacheDisabled: true, ReplicationFactor: numZones},
+				ringDesc:             ringDesc,
+				ringTokens:           ringDesc.GetTokens(),
+				ringTokensByZone:     ringDesc.getTokensByZone(),
+				ringInstanceByToken:  ringDesc.getTokensInfo(),
+				ringZones:            getZones(ringDesc.getTokensByZone()),
+				shuffledSubringCache: map[subringCacheKey]*Ring{},
+				strategy:             NewDefaultReplicationStrategy(),
+				lastTopologyChange:   time.Now(),
+			}
+
+			for id, exp := range testData.expected {
+				ranges, err := ring.GetTokenRangesForInstance(id)
+				require.NoError(t, err)
+				assert.Equal(t, exp, ranges)
+
+				// validate that the endpoints of the ranges map to the expected instances
+				for _, token := range ranges {
+					zoneTokens := ring.ringTokensByZone["zone-0"]
+					i := searchToken(zoneTokens, token)
+					assert.Equal(t, id, ring.ringInstanceByToken[zoneTokens[i]].InstanceID)
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkGetTokenRangesForInstance(b *testing.B) {
+	instancesPerZone := []int{1, 3, 9, 27, 81, 243, 729}
+
+	for _, n := range instancesPerZone {
+		b.Run(fmt.Sprintf("%d_instancesperzone", n), func(b *testing.B) {
+			benchmarkGetTokenRangesForInstance(b, n)
+		})
+	}
+}
+
+func benchmarkGetTokenRangesForInstance(b *testing.B, instancesPerZone int) {
+	numZones := 3
+	numTokens := 512
+
+	b.ResetTimer()
+
+	for n := 0; n < b.N; n++ {
+		b.StopTimer()
+		// Initialise the ring.
+		ringDesc := &Desc{Ingesters: generateRingInstances(instancesPerZone*numZones, numZones, numTokens)}
+		ring := Ring{
+			cfg:                  Config{HeartbeatTimeout: time.Hour, ZoneAwarenessEnabled: true, SubringCacheDisabled: true, ReplicationFactor: numZones},
+			ringDesc:             ringDesc,
+			ringTokens:           ringDesc.GetTokens(),
+			ringTokensByZone:     ringDesc.getTokensByZone(),
+			ringInstanceByToken:  ringDesc.getTokensInfo(),
+			ringZones:            getZones(ringDesc.getTokensByZone()),
+			shuffledSubringCache: map[subringCacheKey]*Ring{},
+			strategy:             NewDefaultReplicationStrategy(),
+			lastTopologyChange:   time.Now(),
+		}
+		b.StartTimer()
+
+		_, _ = ring.GetTokenRangesForInstance("instance-1")
+	}
+}
+
+func TestCheckingOfKeyOwnership(t *testing.T) {
+	const instancesPerZone = 100
+	const numZones = 3
+	const numTokens = 512
+	const replicationFactor = numZones // This is the only config supported by GetTokenRangesForInstance right now.
+
```

Comment:
[nit] You may want to init and log the rand seed here to be able to eventually reproduce a failure.

Reply:
I've added `initTokenGenerator` and passing it to `generateRingInstances` now, modified many tests in the progress. (But not all, there's still `GenerateTokens` function that generates random tokens without logging the seed)

Reply:
Fixed that in other tests too: https://github.com/grafana/dskit/pull/437